### PR TITLE
[Feature] Support column mapping for delta lake

### LINKED ISSUE #51314: Support column mapping for delta lake
- [x] https://github.com/StarRocks/starrocks/pull/51807

## Feature request

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Youngwb**
## Why I'm doing:
Support column mapping for delta lake
### Delta lake Column Mapping
Delta lake can use column mapping to avoid any column naming restrictions, and to support the renaming and dropping of columns without having to rewrite all the data. refer to https://github.com/delta-io/delta/blob/master/PROTOCOL.md#column-mapping for more details.

### Implementation
In id  mode, readers must resolve columns by using the field_id in the parquet metadata for each file
In name mode, readers must resolve columns in the data files by their physical names 
so add physical schema to support both mode:

```
struct TPhysicalSchemaField {
    // field id in parquet schema
    1: optional i64 field_id
    // logical field name
    2: optional string logical_name
    // physical field name
    3: optional string physical_name
    // Children fields for struct, map and list(array)
    10: optional list<TPhysicalSchemaField> children
}

struct TPhysicalSchema {
  1: optional list<TPhysicalSchemaField> fields
}
```
In id  mode, FE set logical_name and field_id,  In name mode, FE set logical_name and physical_name.
In BE, StarRocks **convert field logical name to field_id or physical name** according to the mode, and use field_id or physical_name to read parquet

Fixes #51314

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by Youngwb** [comment]
@mergify rebase

File: gensrc/thrift/Descriptors.thrift [hunk 1]
```diff
@@ -487,6 +487,21 @@ struct TIcebergSchema {
     1: optional list<TIcebergSchemaField> fields
 }
 
+struct TPhysicalSchemaField {
+    // field id in parquet schema
+    1: optional i64 field_id
```
[the remark this answers is not in the document]
Comment (Youngwb) [reply gensrc/thrift/Descriptors.thrift:492]: I want to use set field_id to represent id mode, set physical name to represent name mode,  and the iceberg can  also use this TPhysicalSchemaField to replace the TicebergSchemaField, and there is no mode concept for iceberg

File: be/src/formats/parquet/column_reader.cpp [hunk 2]
```diff
@@ -134,28 +135,87 @@ Status ColumnDictFilterContext::rewrite_conjunct_ctxs_to_predicate(StoredColumnR
...
 
-    // build tmp mapping for ParquetField
-    std::unordered_map<std::string, size_t> field_name_2_pos;
-    for (size_t i = 0; i < field.children.size(); i++) {
-        const std::string format_field_name =
-                case_sensitive ? field.children[i].name : boost::algorithm::to_lower_copy(field.children[i].name);
-        field_name_2_pos.emplace(format_field_name, i);
-    }
+        std::unordered_map<int32_t, size_t> field_id_2_pos;
+        std::unordered_map<std::string, size_t> physical_name_2_pos;
```
Comment (zombee0) [inline be/src/formats/parquet/column_reader.cpp:150]: why not just use logical_name->pos map?

File: be/src/formats/parquet/meta_helper.cpp [hunk 3]
```diff
@@ -81,28 +145,100 @@ bool ParquetMetaHelper::_is_valid_type(const ParquetField* parquet_field, const
...
             }
         }
     } else if (parquet_field->type.is_struct_type()) {
-        std::unordered_map<std::string, const TypeDescriptor*> field_name_2_type{};
-        for (size_t idx = 0; idx < type_descriptor->children.size(); idx++) {
-            field_name_2_type.emplace(Utils::format_name(type_descriptor->field_names[idx], _case_sensitive),
-                                      &type_descriptor->children[idx]);
-        }
+        if (physical_schema_field != nullptr) {
+            if (physical_schema_field->__isset.field_id) {
```
Comment (zombee0) [inline be/src/formats/parquet/meta_helper.cpp:157]: Do you think if we have a logical_name->physical_schema map, we can deal filed id mapping and name mapping in the same logic?

**Review by zombee0 — COMMENTED** [review]
Left LGTM,
 and i think maybe we need to refactor the reader code, in fact the column reader of group reader shouldn't know anything about nested column, and so that we don't need deal nested column both in meta_helper and column_reader.

**Comment by dirtysalt** [comment]
I have a general proposal @zombee0 

Could we abstract a class to maintain our table schema, from parquet schema? So we can
- create parquet schema -> our table schema
- and do field-reordering and field-renaming and field-name-case-sensitive-transition on our table schema
- and column readers depend on our table schema instead of parquet schema

File: be/src/formats/parquet/column_reader.cpp (hunk 2 above)
Comment (Youngwb) [reply be/src/formats/parquet/column_reader.cpp:150]: we need to check the filed_id or physical name exist in parquet field

**Review by zombee0 — APPROVED** [review]

**Review by dirtysalt — DISMISSED** [review]

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/delta/DeltaUtils.java [hunk 4]
```diff
@@ -113,4 +110,54 @@ public static RemoteFileInputFormat getRemoteFileFormat(String format) {
...
+
+    public static TPhysicalSchema getPhysicalSchema(StructType physicalSchema, Metadata metadata) {
+        String columnMappingMode = ColumnMapping.getColumnMappingMode(metadata.getConfiguration());
+        if (columnMappingMode.equals(ColumnMapping.COLUMN_MAPPING_MODE_ID) ||
+                columnMappingMode.equals(ColumnMapping.COLUMN_MAPPING_MODE_NAME)) {
+            TPhysicalSchema tPhysicalSchema = new TPhysicalSchema();
+            for (StructField field : physicalSchema.fields()) {
+                tPhysicalSchema.addToFields(getPhysicalSchemaField(field, columnMappingMode));
+            }
+            return tPhysicalSchema;
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/delta/DeltaUtils.java:122]: Is there a case that meets the following conditions?
1. column mapping mode is id or name
2. the TPhysicalSchemaField from getPhysicalSchemaField() has no physical name or field id

**Pull request closed without merging**